core/src/main/java/com/linecorp/armeria/server/RoutingPredicate.java

```diff
@@ -0,0 +1,227 @@
+        final Predicate<HttpHeaders> predicate =
+                headers -> headers.getAll(parsed.name).stream()
+                                  .anyMatch(valueMatcher(parsed.comparator, parsed.value));
+        return new RoutingPredicate<>(headersPredicate, predicate);
+    }
+
+    /**
+     * Returns a new {@link RoutingPredicate} for {@link HttpParameters}.
+     */
+    static RoutingPredicate<HttpParameters> ofParams(String paramsPredicate) {
+        requireNonNull(paramsPredicate, "paramsPredicate");
+        final Matcher m = CONTAIN_PATTERN.matcher(paramsPredicate);
+        if (m.matches()) {
+            final Predicate<HttpParameters> predicate = params -> params.contains(m.group(2));
+            return new RoutingPredicate<>(paramsPredicate,
+                                          "!".equals(m.group(1)) ? predicate.negate() : predicate);
+        }
+
+        final ParsedComparingPredicate parsed = parse(paramsPredicate);
+        final Predicate<HttpParameters> predicate =
+                params -> params.getAll(parsed.name).stream()
+                                .anyMatch(valueMatcher(parsed.comparator, parsed.value));
+        return new RoutingPredicate<>(paramsPredicate, predicate);
+    }
+
+    private static Predicate<String> valueMatcher(String comparator, String configuredValue) {
+        try {
+            final long longConfiguredValue = Long.parseLong(configuredValue);
+            switch (comparator) {
+                case "=":
+                    return longValueMatcher(value -> value == longConfiguredValue);
+                case "!=":
+                    return longValueMatcher(value -> value != longConfiguredValue);
+                case ">":
+                    return longValueMatcher(value -> value > longConfiguredValue);
+                case ">=":
+                    return longValueMatcher(value -> value >= longConfiguredValue);
+                case "<":
+                    return longValueMatcher(value -> value < longConfiguredValue);
+                case "<=":
+                    return longValueMatcher(value -> value <= longConfiguredValue);
+            }
+        } catch (NumberFormatException nfe) {
+            switch (comparator) {
+                case "=":
+                    return configuredValue::equals;
+                case "!=":
+                    return value -> !value.equals(configuredValue);
+                case ">":
+                    return value -> value.compareTo(configuredValue) > 0;
+                case ">=":
+                    return value -> value.compareTo(configuredValue) >= 0;
+                case "<":
+                    return value -> value.compareTo(configuredValue) < 0;
+                case "<=":
+                    return value -> value.compareTo(configuredValue) <= 0;
+            }
+        }
+        throw new IllegalStateException("Pattern has changed unexpectedly: " + COMPARE_PATTERN.pattern());
```

Comment:
What does this mean? How about rephrasing the exception message so it's more user-friendly?

Reply:
Removed. :-)